examples/src/main/java/com/squareup/moshi/recipes/RequiredJsonAdapterFactory.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class RequiredJsonAdapterFactory implements JsonAdapter.Factory {
+  private static final class Train {
+    @Required String name;
+  }
+
+  public static void main(String[] args) {
+    Moshi moshi = new Moshi.Builder().add(new RequiredJsonAdapterFactory()).build();
+    JsonAdapter<Train> adapter = moshi.adapter(Train.class);
+    try {
+      adapter.fromJson("{}");
+      throw new AssertionError();
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    } catch (IllegalStateException expected) {
+      System.err.println(expected.getMessage());
+    }
+  }
+
+  @Retention(RUNTIME) @interface Required {
+  }
+
+  @Override
+  public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+    final JsonAdapter<Object> delegate = moshi.nextAdapter(this, type, annotations);
+    return new JsonAdapter<Object>() {
+      @Override public Object fromJson(JsonReader reader) throws IOException {
+        Object value = delegate.fromJson(reader);
+        if (value == null) {
+          return null;
+        }
+        checkRequiredFields(value);
+        return value;
+      }
+
+      @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+        if (value == null) {
+          writer.nullValue();
+          return;
+        }
+        checkRequiredFields(value);
+        delegate.toJson(writer, value);
+      }
+
+      private void checkRequiredFields(Object value) {
```

Comment:
Better would be to precompute the `List<Field>` that are required, and only if non-empty do you even return an adapter from the factory